File: client/js/src/instructions/update-player-authority.ts
import {
  IAccountLookupMeta,
  IAccountMeta,
  IInstruction,
} from "@solana/instructions";
import { ReadonlyUint8Array, Address, AccountRole } from "gill";
import { getPlayerAddress } from "../accounts";
import { TOWER_DEFENSE_PROGRAM_ID } from "../constants";
import { InstructionDiscriminator } from ".";

export interface UpdatePlayerAuthorityInstructionArgs {
  signer: Address;
  newAuthority: Address;
}

export class UpdatePlayerAuthority implements IInstruction {
  accounts:
    | readonly (IAccountLookupMeta<string, string> | IAccountMeta<string>)[]
    | undefined;

  data: ReadonlyUint8Array | undefined;

  programAddress: Address<string> = TOWER_DEFENSE_PROGRAM_ID;

  args: UpdatePlayerAuthorityInstructionArgs;

  constructor(args: UpdatePlayerAuthorityInstructionArgs) {
    this.args = args;
  }

  public async make() {
    const { signer, newAuthority } = this.args;
    const { player } = await getPlayerAddress(signer);

    const allBytes = [InstructionDiscriminator.UpdatePlayerAuthority];

    const data: ReadonlyUint8Array = Uint8Array.from(allBytes);

    const accounts = [
      { address: player, addressIndex: 0, role: AccountRole.WRITABLE },
      {
        address: signer,
        addressIndex: 1,
        role: AccountRole.WRITABLE_SIGNER,
      },
      {
        address: newAuthority,
        addressIndex: 2,
        role: AccountRole.READONLY,
      },
    ];
    this.accounts = accounts;
    this.data = data;
  }
}

File: client/rust/src/instructions/update_player_authority.rs
use solana_sdk::{
    instruction::{AccountMeta, Instruction},
    pubkey::Pubkey,
};

use crate::PROGRAM_ID;

use super::InstructionDiscriminator;

#[inline(always)]
pub fn update_player_authority(
    player: &Pubkey,
    signer: &Pubkey,
    new_authority: &Pubkey,
) -> Instruction {
    let account_metas: [AccountMeta; 3] = [
        AccountMeta::new(*player, false),
        AccountMeta::new(*signer, true),
        AccountMeta::new(*new_authority, false),
    ];

    // Instruction data layout:
    // -  [0]: instruction discriminator (1 byte, u8)
    let data_len = 1;
    let instruction_data = vec![InstructionDiscriminator::UpdatePlayerAuthority as u8; data_len];

    Instruction {
        program_id: PROGRAM_ID,
        accounts: account_metas.to_vec(),
        data: instruction_data,
    }
}
